File: apps/backends/node/src/services/checkout.ts
import { Pool } from 'pg';
import { v4 as uuidv4 } from 'uuid';
import { getProduct, insertOrder } from '../db/queries.js';
import { computeFraudScore } from './fraud.js';
import { calculateTax } from '../clients/tax.js';
import { CheckoutRequest, CheckoutResponse, OrderItem } from '../types.js';

export async function processCheckout(
  req: CheckoutRequest,
  pool: Pool,
  taxServiceUrl: string
): Promise<CheckoutResponse> {
  if (!req.items || req.items.length === 0 || req.items.length > 8) {
    throw new Error('Cart must have 1-8 items');
  }

  const uuidRegex = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;
  if (!uuidRegex.test(req.customer_id)) {
    throw new Error('Invalid customer ID');
  }

  let subtotal = 0;
  const orderItems: OrderItem[] = [];

  for (const item of req.items) {
    if (!uuidRegex.test(item.product_id)) {
      throw new Error('Invalid product ID');
    }

    const product = await getProduct(pool, item.product_id);
    if (product.stock < item.quantity) {
      throw new Error('Insufficient stock');
    }

    subtotal += product.price_cents * item.quantity;
    orderItems.push({
      product_id: item.product_id,
      quantity: item.quantity,
      price_cents: product.price_cents,
    });
  }

  const taxResp = await calculateTax(taxServiceUrl, subtotal, req.state);
  const fraudScore = computeFraudScore(subtotal, orderItems.length);
  const orderId = uuidv4();
  const total = subtotal + taxResp.tax_cents;

  await insertOrder(pool, { 
    id: orderId, 
    customer_id: req.customer_id, 
    total_cents: total, 
    tax_cents: taxResp.tax_cents 
  }, orderItems);

  return {
    order_id: orderId,
    total_cents: total,
    tax_cents: taxResp.tax_cents,
    fraud_score: fraudScore,
  };
}

File: apps/backends/python/app/services/checkout.py
import asyncpg
import uuid as uuid_module
from app.models.types import CheckoutRequest, CheckoutResponse, OrderItem
from app.db.queries import get_product, insert_order
from app.services.fraud import compute_fraud_score
from app.clients.tax import calculate_tax

async def process_checkout(
    req: CheckoutRequest,
    pool: asyncpg.Pool,
    tax_service_url: str,
) -> CheckoutResponse:
    if not req.items or len(req.items) == 0 or len(req.items) > 8:
        raise ValueError('Cart must have 1-8 items')

    # Validate customer ID (UUID format)
    try:
        customer_uuid = uuid_module.UUID(req.customer_id)
    except (ValueError, TypeError):
        raise ValueError('Invalid customer ID')

    subtotal = 0
    order_items = []

    for item in req.items:
        try:
            product_id = uuid_module.UUID(item.product_id)
        except (ValueError, TypeError):
            raise ValueError('Invalid product ID')

        product = await get_product(pool, item.product_id)
        if product['stock'] < item.quantity:
            raise ValueError('Insufficient stock')

        subtotal += product['price_cents'] * item.quantity
        order_items.append(OrderItem(
            product_id=item.product_id,
            quantity=item.quantity,
            price_cents=product['price_cents'],
        ))

    tax_resp = await calculate_tax(tax_service_url, subtotal, req.state)
    fraud_score = compute_fraud_score(subtotal, len(order_items))
    order_id = str(uuid_module.uuid4())
    total = subtotal + tax_resp.tax_cents

    await insert_order(
        pool,
        order_id,
        req.customer_id,
        total,
        tax_resp.tax_cents,
        [dict(
            product_id=item.product_id,
            quantity=item.quantity,
            price_cents=item.price_cents,
        ) for item in order_items],
    )

    return CheckoutResponse(
        order_id=order_id,
        total_cents=total,
        tax_cents=tax_resp.tax_cents,
        fraud_score=fraud_score,
    )
